# Python Functions

In this class, you'll learn about functions, what a function is, the syntax, components, and types of functions. Also, you'll learn to create a function in Python.

# What is a function in Python?

In Python, a **function is a block of organized, reusable (DRY- Don’t Repeat Yourself) code with a name** that is used to perform a single, specific task. It can take arguments and returns the value.

Functions help break our program into smaller and modular chunks. As our program grows larger and larger, functions make it more organized and manageable.

Furthermore, it improves efficiency and reduces errors because of the reusability of a code.

**Syntax:**

```python
def function_name(parameter1, parameter2):
    """docstring"""
    # function body    
    # write some action
return value
```

<div>
<img src="../img/f1.png" width="550"/>
</div>

## Defining a function without any parameters

Function can be declared without parameters.

In [1]:
# Example 1: 
    
def greet():
    print("Welcome to Python for Data Science")

# call function using its name
greet()

Welcome to Python for Data Science


In [2]:
# Example 2: 

def add_two_numbers ():
    num_one = 3
    num_two = 6
    total = num_one + num_two
    print(total)
add_two_numbers() # calling a function

9


In [3]:
# Example 3: 

def generate_full_name ():
    first_name = 'Milaan'
    last_name = 'Parmar'
    space = ' '
    full_name = first_name + space + last_name
    print(full_name)
generate_full_name () # calling a function

Milaan Parmar


## Defining a function without parameters and `return` value

Function can also return values, if a function does not have a **`return`** statement, the value of the function is None. Let us rewrite the above functions using **`return`**. From now on, we get a value from a function when we call the function and print it.

In [4]:
# Example 1: 

def add_two_numbers ():
    num_one = 3
    num_two = 6
    total = num_one + num_two
    return total
print(add_two_numbers())

9


In [5]:
# Example 2: 

def generate_full_name ():
    first_name = 'Milaan'
    last_name = 'Parmar'
    space = ' '
    full_name = first_name + space + last_name
    return full_name
print(generate_full_name())

Milaan Parmar


## Defining a function with parameters

In a function we can pass different data types(number, string, boolean, list, tuple, dictionary or set) as a parameter.

### Single Parameter: 

If our function takes a parameter we should call our function with an argument

In [6]:
# Example 1: Gereeting

def greet(name):
    """
    This function greets to the person passed in as a parameter
    """
    print("Hello, " + name + ". Good morning!")   # No output!

In [7]:
# Example 2: 

def sum_of_numbers(n):
    total = 0
    for i in range(n+1):
        total+=i
    print(total)
print(sum_of_numbers(10))  # 55
print(sum_of_numbers(100)) # 5050

55
None
5050
None


### Two Parameter: 

A function may or may not have a parameter or parameters. A function may also have two or more parameters. If our function takes parameters we should call it with arguments.

In [8]:
# Example 1: 

def course(name, course_name):
    print("Hello", name, "Welcome to Python for Data Science")
    print("Your course name is", course_name)

course('Arthur', 'Python')   # call function

Hello Arthur Welcome to Python for Data Science
Your course name is Python


## Defining a function with parameters and `return` value

In [9]:
# Example 1: 

def greetings (name):  # single parameter
    message = name + ', welcome to Python for Data Science'
    return message

print(greetings('Milaan'))

Milaan, welcome to Python for Data Science


In [10]:
# Example 2: 

def add_ten(num):  # single parameter
    ten = 10
    return num + ten
print(add_ten(90))

100


In [11]:
# Example 3: 

def square_number(x):  # single parameter
    return x * x
print(square_number(3))

9


In [12]:
# Example 4: 

def area_of_circle (r):  # single parameter
    PI = 3.14
    area = PI * r ** 2
    return area
print(area_of_circle(10))

314.0


In [13]:
# Example 5: 

def calculator(a, b):  # two parameter
    add = a + b   
    return add    # return the addition

result = calculator(30, 6)   # call function & take return value in variable
print("Addition :", result)   # Output Addition : 36

Addition : 36


In [14]:
# Example 6: 

def generate_full_name (first_name, last_name):  # two parameter
    space = ' '
    full_name = first_name + space + last_name
    return full_name
print('Full Name: ', generate_full_name('Milaan','Parmar'))

Full Name:  Milaan Parmar


In [15]:
# Example 7: 

def sum_two_numbers (num_one, num_two):  # two parameter
    sum = num_one + num_two
    return sum
print('Sum of two numbers: ', sum_two_numbers(1, 9))

Sum of two numbers:  10


In [16]:
# Example 8: 

def calculate_age (current_year, birth_year):  # two parameter
    age = current_year - birth_year
    return age;

print('Age: ', calculate_age(2021, 1819))

Age:  202


In [17]:
# Example 9: 

def weight_of_object (mass, gravity):  # two parameter
    weight = str(mass * gravity)+ ' N' # the value has to be changed to a string first
    return weight
print('Weight of an object in Newtons: ', weight_of_object(100, 9.81))

Weight of an object in Newtons:  981.0 N


## Function `return` Statement

In Python, to return value from the function, a **`return`** statement is used. It returns the value of the expression following the returns keyword.

**Syntax:**

```python
def fun():
    statement-1
    statement-2
    statement-3
    .          
    .          
    return [expression]
```

The **`return`** value is nothing but a outcome of function.

* The **`return`** statement ends the function execution.
* For a function, it is not mandatory to return a value.
* If a **`return`** statement is used without any expression, then the **`None`** is returned.
* The **`return`** statement should be inside of the function block.

### Return Single Value

In [18]:
print(greet("Cory"))

Hello, Cory. Good morning!
None


Here, **`None`** is the returned value since **`greet()`** directly prints the name and no **`return`** statement is used.

## Passing Arguments with Key and Value

If we pass the arguments with key and value, the order of the arguments does not matter.

In [19]:
# Example 1: 

def print_fullname(firstname, lastname):
    space = ' '
    full_name = firstname  + space + lastname
    print(full_name)
print(print_fullname(firstname = 'Milaan', lastname = 'Parmar'))

Milaan Parmar
None


In [20]:
# Example 2:

def add_two_numbers (num1, num2):
    total = num1 + num2
    print(total)
print(add_two_numbers(num2 = 3, num1 = 2)) # Order does not matter

5
None


If we do not **`return`** a value with a function, then our function is returning **`None`** by default. To return a value with a function we use the keyword **`return`** followed by the variable we are returning. We can return any kind of data types from a function.

In [21]:
# Example 1:  with return statement

def print_fullname(firstname, lastname):
    space = ' '
    full_name = firstname  + space + lastname
    return full_name
print(print_fullname(firstname = 'Milaan', lastname = 'Parmar'))

Milaan Parmar


In [22]:
# Example 2:  with return statement

def add_two_numbers (num1, num2):
    total = num1 + num2
    return total
print(add_two_numbers(num2 = 3, num1 = 2)) # Order does not matter

5


In [23]:
# Example 3:

def absolute_value(num):
    """This function returns the absolute
    value of the entered number"""

    if num >= 0:
        return num
    else:
        return -num

print(absolute_value(2))
print(absolute_value(-4))

2
4


In [24]:
# Example 4:

def sum(a,b):  # Function 1
    print("Adding the two values")
    print("Printing within Function")
    print(a+b)
    return a+b

def msg():  # Function 2
    print("Hello")
    return

total=sum(10,20)
print('total : ',total)
msg()
print("Rest of code")

Adding the two values
Printing within Function
30
total :  30
Hello
Rest of code


In [25]:
# Example 5:

def is_even(list1):
    even_num = []
    for n in list1:
        if n % 2 == 0:
            even_num.append(n)
    # return a list
    return even_num

# Pass list to the function
even_num = is_even([2, 3, 46, 63, 72, 83, 90, 19])
print("Even numbers are:", even_num)

Even numbers are: [2, 46, 72, 90]


### Return Multiple Values

You can also return multiple values from a function. Use the return statement by separating each expression by a comma.

In [26]:
# Example 1:

def arithmetic(num1, num2):
    add = num1 + num2
    sub = num1 - num2
    multiply = num1 * num2
    division = num1 / num2
    # return four values
    return add, sub, multiply, division

a, b, c, d = arithmetic(10, 2)  # read four return values in four variables

print("Addition: ", a)
print("Subtraction: ", b)
print("Multiplication: ", c)
print("Division: ", d)

Addition:  12
Subtraction:  8
Multiplication:  20
Division:  5.0


### Return Boolean Values

In [27]:
# Example 1:

def is_even (n):
    if n % 2 == 0:
        print('even')
        return True    # return stops further execution of the function, similar to break 
    return False
print(is_even(10)) # True
print(is_even(7)) # False

even
True
False


### Return a List

In [28]:
# Example 1:

def find_even_numbers(n):
    evens = []
    for i in range(n + 1):
        if i % 2 == 0:
            evens.append(i)
    return evens
print(find_even_numbers(10))

[0, 2, 4, 6, 8, 10]


## How to call a function in python?

Once we have defined a function, we can call it from another function, program or even the Python prompt. To call a function we simply type the function name with appropriate parameters.

<div>
<img src="../img/f2.png" width="500"/>
</div>

In [29]:
greet('Alan')

Hello, Alan. Good morning!


>**Note:** Try running the above code in the Python program with the function definition to see the output.

In [30]:
# Example 1: 

def wish(name):
    """
    This function wishes to the person passed in as a parameter
    """
    print("Happy birthday, " + name + ". Hope you have a wonderful day!")

wish('Bill')

Happy birthday, Bill. Hope you have a wonderful day!


In [31]:
# Example 2: 

def greetings (name = 'Clark'):
    message = name + ', welcome to Python for Data Science'
    return message
print(greetings())
print(greetings('Milaan'))

Clark, welcome to Python for Data Science
Milaan, welcome to Python for Data Science


In [32]:
# Example 3: 

def generate_full_name (first_name = 'Milaan', last_name = 'Parmar'):
    space = ' '
    full_name = first_name + space + last_name
    return full_name
print(generate_full_name())
print(generate_full_name('Ethan','Hunt'))

Milaan Parmar
Ethan Hunt


In [33]:
# Example 4: 

def calculate_age (birth_year,current_year = 2021):
    age = current_year - birth_year
    return age;
print('Age: ', calculate_age(1821))

Age:  200


In [34]:
# Example 5: 

def swap(x, y):
    """
    This function swaps the value of two variables
    """
    temp = x;  # value of x will go inside temp
    x = y;     # value of y will go inside x
    y = temp;  # value of temp will go inside y
    print("value of x is:", x)
    print("value of y is:", y)
    return     # "return" is optional

x = 6
y = 9
swap(x, y)     #call function

value of x is: 9
value of y is: 6


In [35]:
# Example 6: 

def even_odd(n):    
    if n % 2 == 0:   # check number is even or odd
        print(n, 'is a Even number')
    else:
        print(n, 'is a Odd Number')

even_odd(9)   # calling function by its name

9 is a Odd Number


In [36]:
# Example 7: 

def weight_of_object (mass, gravity = 9.81):
    weight = str(mass * gravity)+ ' N' # the value has to be changed to string first
    return weight
print('Weight of an object in Newtons: ', weight_of_object(100)) # 9.81 - average gravity on Earth's surface
print('Weight of an object in Newtons: ', weight_of_object(100, 1.62)) # gravity on the surface of the Moon

Weight of an object in Newtons:  981.0 N
Weight of an object in Newtons:  162.0 N



## Docstrings

The first string after the function header is called the **docstring** and is short for documentation string. It is a descriptive text (like a comment) written by a programmer to let others know what block of code does.

Although **optional**, documentation is a good programming practice. Unless you can remember what you had for dinner last week, always document your code.

It is being declared using triple single quotes **`''' '''`** or triple-double quote **`""" """`** so that docstring can extend up to multiple lines.

We can access docstring using doc attribute **`__doc__`** for any object like list, tuple, dict, and user-defined function, etc.

In the above example, we have a docstring immediately below the function header.

In [37]:
print(greet.__doc__)


    This function greets to the person passed in as a parameter
    


## Function `pass` Statement

In Python, the **`pass`** is the keyword, which won’t do anything. Sometimes there is a situation where we need to define a syntactically empty block. We can define that block using the **`pass`** keyword.

When the interpreter finds a **`pass`** statement in the program, it returns no operation.

In [38]:
# Example 1:

def addition(num1, num2):
    # Implementation of addition function in comming release
    # Pass statement 
    pass

addition(10, 2)

# Python Global, Local and Nonlocal variables

In this class, you’ll learn about Python Global variables, Local variables, Nonlocal variables and where to use them.

When we define a function with variables, then those variables scope is limited to that function. In Python, the scope of a variable is the portion of a program where the variable is declared. Parameters and variables defined inside a function are not visible from outside the function. Hence, it is called the variable’s local scope.

>**Note:** The inner function does have access to the outer function’s local scope.

When we are executing a function, the life of the variables is up to running time. Once we return from the function, those variables get destroyed. So function does no need to remember the value of a variable from its previous call.

## Global Variables

In Python, a variable declared outside of the function or in global scope is known as a global variable. This means that a global variable can be accessed inside or outside of the function.

For example:

In [1]:
# Example 1: Create a Global Variable

global_var = 999

def fun1():
    print("Value in 1st function:", global_var)

def fun2():
    print("Value in 2nd function:", global_var)

fun1()
fun2()

Value in 1st function: 999
Value in 2nd function: 999


In [2]:
# Example 2: 

x = "global"

def fun():
    print("x inside:", x)

fun()
print("x outside:", x)

x inside: global
x outside: global


In the above code, we created **`x`** as a global variable and defined a **`fun()`** to print the global variable **`x`**. Finally, we call the **`fun()`** which will print the value of **`x`**.

What if you want to change the value of **`x`** inside a function?

In [3]:
# Example 3: 

x = "global"

def fun():
    x = x * 2
    print(x)

fun()

UnboundLocalError: local variable 'x' referenced before assignment

In [4]:
# Example 4: 

global_lang = 'DataScience'

def var_scope_test():
    local_lang = 'Python'
    print(local_lang)

var_scope_test()  # Output 'Python'

# outside of function
print(global_lang)   # Output 'DataScience'

print(local_lang)   # NameError: name 'local_lang' is not defined

Python
DataScience


NameError: name 'local_lang' is not defined

In [5]:
# Example 5: 

a=90   # 'a' is a variable defined outside of function, i.e., Global variable

def print_data():
    a=6  # 'a' is a variable defined inside of function, i.e., local variable
    b=30
    print("(a,b):(",a,",",b,")")

print_data() #(a,b):( 5 , 10 )
print("Global a :",a)  #Global x : 50
print("Local b : ",b)  #b is local veriable - throw NameError

(a,b):( 6 , 30 )
Global a : 90


NameError: name 'b' is not defined

## Local Variables

A variable declared inside the function's body or in the local scope is known as a local variable.

If we try to access the local variable from the outside of the function, we will get the error as **`NameError`**.

In [6]:
# Example 1: Accessing local variable outside the scope

def fun():
    y = "local"
    
fun()
print(y)

NameError: name 'y' is not defined

The output shows an error because we are trying to access a local variable **`y`** in a global scope whereas the local variable only works inside **`fun()`** or local scope.

In [7]:
# Example 2: Create a Local Variable

# Normally, we declare a variable inside the function to create a local variable.

def fun():
    y = "local"
    print(y)

fun()

local


Let's take a look at the cell **In [2]: # Example 3:** where **`x`** was a global variable and we wanted to modify **`x`** inside **`fun()`**.

In [8]:
# Exercise 3: 

def fun1():
    loc_var = 999   # local variable
    print("Value is :", loc_var)

def fun2():
    print("Value is :", loc_var)

fun1()
fun2()

Value is : 999


NameError: name 'loc_var' is not defined

## Global and local variables

Here, we will show how to use global variables and local variables in the same code.

In [9]:
# Example 1: Using Global and Local variables in the same code

x = "global"

def fun():
    global x
    y = "local"
    x = x * 2
    print(x)
    print(y)

fun()

globalglobal
local


**Explanation**:

In the above code, we declare **`x`** as a global and **`y`** as a local variable in the **`fun()`**. Then, we use multiplication operator **`*`** to modify the global variable **`x`** and we print both **`x`** and **`y`**.

After calling the **`fun()`**, the value of **`x`** becomes **`global global`** because we used the **`x * 2`** to print two times **`global`**. After that, we print the value of local variable **`y`** i.e **`local`**.

In [10]:
# Example 2: Global variable and Local variable with same name

x = 9

def fun():
    x = 19
    print("local x:", x)


fun()
print("global x:", x)

local x: 19
global x: 9


**Explanation**:

In the above code, we used the same name **`x`** for both global variable and local variable. We get a different result when we print the same variable because the variable is declared in both scopes, i.e. the local scope inside **`fun()`** and global scope outside **`fun()`**.

When we print the variable inside **`fun()`** it outputs **`local x: 19`**. This is called the local scope of the variable.

Similarly, when we print the variable outside the **`fun()`**, it outputs **`global x: 9`**. This is called the global scope of the variable.

In [11]:
# Exercise 3: 

def my_func():  # for this Function I am not writing any argument in parenthesis '()'
    x = 10
    print("Value inside the body of function:",x)

x = 20                             # first, this line to execute
my_func()                          # second, the body of function will execute
print("Value outside of function:",x) # finally, this line will execute

Value inside the body of function: 10
Value outside of function: 20


**Explanation:**

Here, we can see that the value of **`x`** is 20 initially. Even though the function **`my_func()`** changed the value of **`x`** to 10, it did not affect the value outside the function.

This is because the variable **`x`** inside the function is different (local to the function) from the one outside. Although they have the same names, they are two different variables with different scopes.

On the other hand, variables outside of the function are visible from inside. They have a global scope.

We can read these values from inside the function but cannot change (write) them. In order to modify the value of variables outside the function, they must be declared as global variables using the keyword global.

## Nonlocal Variables

Nonlocal variables are used in nested functions whose local scope is not defined. This means that the variable can be neither in the local nor the global scope.

Let's see an example of how a global variable is created in Python.

We use **`nonlocal`** keywords to create nonlocal variables.

In [12]:
# Example 1: Create a nonlocal variable

x1 = "global"  # Global variable

def outer_fun():  # main function
    x1 = "local"  # 'x' is local variable for main function and it is nested variable for nested function
    print("variable type for Outer function:", x1)
    
    def inner_fun():  # nested fucntion
        nonlocal x1   # using local variable 'x' in nested function as nonloval variable
        x1 = "nonlocal"  # changing the value of my 'x'
        print("variable type for Inner function:", x1)  # print 'nonlocal'

    inner_fun()   #print("outer:", x1)  # print 'nonlocal'
    
outer_fun()
print("Variable type of x1:", x1)

variable type for Outer function: local
variable type for Inner function: nonlocal
Variable type of x1: global


In the above code, there is a nested **`inner()`** function. We use nonlocal keywords to create a **`nonlocal`** variable. The **`inner()`** function is defined in the scope of another function **`outer()`**.

> **Note**: If we change the value of a nonlocal variable, the changes appear in the local variable.

In [13]:
# Exercise 2: 

def outer_fun():
    x = 999

    def inner_fun():
        # local variable now acts as global variable
        nonlocal x
        x = 900
        print("value of x inside inner function is:", x)

    inner_fun()
    print("value of x inside outer function is:", x)

outer_fun()

value of x inside inner function is: 900
value of x inside outer function is: 900


# What is `global` keyword?

In Python, **`global`** keyword allows you to modify the variable outside of the current scope. It is used to create a global variable and make changes to the variable in a local context.

## Rules of `global` Keyword

The basic rules for **`global`** keyword in Python are:

1. When we create a variable inside a function, it is local by default.
2. When we define a variable outside of a function, it is global by default. You don't have to use **`global`** keyword.
3. We use **`global`** keyword to read and write a global variable inside a function.
4. Use of **`global`** keyword outside a function has no effect.

## Use of `global` Keyword

In [1]:
# Example 1: Accessing global Variable From Inside a Function

a = 1 # global variable

def add():
    print(a)

add()

1


However, we may have some scenarios where we need to modify the global variable from inside a function.

In [2]:
# Example 2: Modifying Global Variable From Inside the Function

a = 1 # global variable
    
def add():
    a = a + 3 # increment a by 3 (modifying my global variable)
    print(a)

add()

UnboundLocalError: local variable 'a' referenced before assignment

This is because we can only access the global variable but cannot modify it from inside the function.

The solution for this is to use the **`global`** keyword.

In [3]:
# Example 2: Changing Global Variable From Inside a Function using global

a = 0 # global variable

def add():
    global a  # using global variable inside my function
    a = a + 3 # increment by 3
    print("Inside add():", a)

add()
print("In main:", a)

Inside add(): 3
In main: 3


**Explanation:**

In the above program, we define **`a`** as a global keyword inside the **`add()`** function.

Then, we increment the variable **`a`** by 1, i.e **`a = a + 3`**. After that, we call the **`add()`** function. Finally, we print the global variable **`a`**.

As we can see, change also occurred on the global variable outside the function, **`a = 3`**.

Let’s see another example where we don’t use **`global`** keyword to access the global variable in the function.

In [4]:
# Exercise 3: without `global` keyword.

global_var = 9   # Global variable

def fun1():
    print("Value in 1st function:", global_var)

def fun2():
    # Modify global variable
    # function will treat it as a local variable
    global_var = 999
    print("Value in 2nd function:", global_var)

def fun3():
    print("Value in 3rd function:", global_var)

fun1()
fun2()
fun3()

Value in 1st function: 9
Value in 2nd function: 999
Value in 3rd function: 9


**Explanation:**

As you can see, **`fun2()`** treated **`global_var`** as a new variable (local variable). To solve such issues or access/modify global variables inside a function, we use the **`global`** keyword.

In [5]:
# Exercise 3: use the `global` keyword.


x = 9   # Global variable

# defining 1st function
def fun1():
    print("Value in 1st function:", x)

# defining 2nd function
def fun2():
    # Modify global variable using global keyword
    global x
    x = 999
    print("Value in 2nd function:", x)

# defining 3rd function
def fun3():
    print("Value in 3rd function:", x)

fun1()
fun2()
fun3()

Value in 1st function: 9
Value in 2nd function: 999
Value in 3rd function: 999


## Global Variables Across Python Modules

In Python, we create a single module **config.py** to hold global variables and share information across Python modules within the same program.

Here is how we can share global variables across the python modules.

### Example : Share a global Variable Across Python Modules

Create a **config.py** file, to store global variables

```python
>>> a = 0
>>> b = "empty"
```

Create a **update.py** file, to change global variables

```python
>>> import config  # import config.py here

>>> config.a = 10  # change the value of 'a' from 0 to 10
>>> config.b = "alphabet"  # change the value of 'b' from "empty" to "alphabet"
```

Create a **main1.py** file, to test changes in value

```python
>>> import config  # import config.py here
>>> import update  # import update.py here

>>> print(config.a)  # print the updated value of 'a'
>>> print(config.b)  # print the updated value of 'b'
```

When we run the **main.py** file, the output will be

`
10
alphabet
`

In the above, we have created three files: **config.py**, **update.py**, and **main.py**.

The module **config.py** stores global variables of **`a`** and **`b`**. In the **update.py** file, we import the **config.py** module and modify the values of **`a`** and **`b`**. Similarly, in the **main.py** file, we import both **config.py** and **update.py** module. Finally, we print and test the values of global variables whether they are changed or not.

## Global in Nested Functions

Here is how you can use a global variable in nested function.

In [6]:
# Example 1: Using a Global Variable in Nested Function

def fun():  # main function
    x = 30

    def day():  # nested fucntion
        global x
        x = 66
    
    print("Before calling bar function:", x)  # check indentation
    print("Calling bar function now")
    day()
    print("After calling bar function:", x)

fun()
print("x in main function:", x)

Before calling bar function: 30
Calling bar function now
After calling bar function: 30
x in main function: 66


**Explanation:**

In the above program, we declared a global variable inside the nested function **`day()`**. Inside **`fun()`** function, **`x`** has no effect of the global keyword.

Before and after calling **`day()`**, the variable **`x`** takes the value of local variable i.e **`x = 30`**. Outside of the **`fun()`** function, the variable **`x`** will take value defined in the **`day()`** function i.e **`x = 66`**. This is because we have used global keyword in **`x`** to create global variable inside the **`day()`** function (local scope).

If we make any changes inside the **`day()`** function, the changes appear outside the local scope, i.e. **`fun()`**.

# Function Argument and Parameter

The argument is a value, a variable, or an object that we pass to a function or method call. 

There can be two types of data passed in the function.

* The First type of data is the data passed in the function call. This data is called **arguments**.

* The second type of data is the data received in the function definition. This data is called **parameters**.

    - Arguments can be literals, variables and expressions. 
    - Parameters must be variable to hold incoming values.

Alternatively, arguments can be called as **actual parameters** or **actual arguments** and parameters can be called as **formal parameters** or **formal arguments**.

# Python Function Arguments

In Python, you can define a function that takes variable number of arguments. In this article, you will learn to define such functions using default, keyword and arbitrary arguments.

In Python, there are 2 types of arguments allowed.
<b>
1. Positional Arguments (Basic)
2. Variable Function Arguments
</b>

In [1]:
# Example 1:

def greet(name, msg):
    """This function greets to the person with the provided message"""
    print("Hello", name + ', ' + msg)

greet("Arthur", "Good morning!")

Hello Arthur, Good morning!


**Explanation:**

Here, the function **`greet()`** has two parameters.

Since we have called this function with two arguments, it runs smoothly and we do not get any error.

If we call it with a different number of arguments, the interpreter will show an error message. Below is a call to this function with one and no arguments along with their respective error messages.

In [2]:
greet("Arthur")    # only one argument

TypeError: greet() missing 1 required positional argument: 'msg'

In [3]:
greet()    # no arguments

TypeError: greet() missing 2 required positional arguments: 'name' and 'msg'

In [4]:
# Example 2:

def add(a, b):
    print(a - b)

add(90, 30)  # Output 60
add(60, 30)  # Output -30

60
30


In [5]:
# Example 2: If you try to use pass more parameters you will get an error.

def add(a, b):
    print(a - b)
add(109, 633, 9)

TypeError: add() takes 2 positional arguments but 3 were given

>**Note:** In the **Positional argument** number and position of arguments must be matched. If we change the order, then the result may change. Also, If we change the number of arguments, then we will get an error.

In [6]:
# Example 3:

def evenodd(x):     # x is parameters or formal parameters or formal arguments.
    if (x % 2 == 0):
        print("even")
    else:
        print("odd")

# Driver code
evenodd(6)   # here 2 is argument or actual perameter
evenodd(9)   # here 3 is argument or actual perameter

even
odd


In [7]:
# Example 4:

def cube(x):
    "This x a passed num value into this function, return cube of x"
    y=x*x*x;
    return y

# Now you can call cube function
z=cube(2)  #required to pass argument
print(z)

8


## Variable Function Arguments

Up until now, functions had a fixed number of arguments. In Python, there are other ways to define a function that can take variable number of arguments.

Three different forms of this type are described below:
<b>
1. Default Arguments
2. Keyword Arguments
3. Arbitrary/Variable-length Arguments
</b>

### Python Default Arguments

Default arguments are arguments that take the default value during the function call. If we do not pass any argument to the function, then the default argument will take place. We can assign default values using the **`=`** assignment operator. For example:

In [8]:
# Example 1:

def greet(name, msg="Good morning!"):  # two arguments: `name` is fixed arg and 'msg' is variable arg

    print("Hello", name + ', ' + msg)


greet("Alan")
greet("Bruce", "How do you do?")
greet("Carson","Good night!")

Hello Alan, Good morning!
Hello Bruce, How do you do?
Hello Carson, Good night!


**Explanation:**

In this function, the parameter **`name`** does not have a default value and is required (mandatory) during a call.

On the other hand, the parameter **`msg`** has a default value of **`"Good morning!"`**. So, it is optional during a call. If a value is provided, it will overwrite the default value.

Any number of arguments in a function can have a default value. But once we have a default argument, all the arguments to its right must also have default values.

This means to say, non-default arguments cannot follow default arguments. For example, if we had defined the function header above as:

```python
>>> def greet(msg = "Good morning!", name):
```

We would get an error as:
```python
SyntaxError: non-default argument follows default argument
```

In [9]:
# Example 2: function with default argument

def message(name="Everyone"):
    print("Hello", name)

# calling function with argument
message("David")

# calling function without argument
message()

Hello David
Hello Everyone


**Explanation:** 

When we call a function with an argument, it will be considered that value.

Like in the above example, we passed **`name="David"`** we pass to the function, then the function will consider that value. If we do not pass any argument, it will be considered **`"name=Everyone"`** as a default value.

In [10]:
# Example 3: it prints default age if it is not passed

def emp_data(name, emp_id, age, company="Baidu.com"):  # total 4 arguments (3 is fixed and 1 is variable)
    print("Details of: ",name)
    print("Empoyee Id : ",emp_id)
    print("Age : ",age)
    print("Company : ",company)

#call emp_data fun
emp_data("Bill",101,21,"Samsung.com")
print("-----------------------")
emp_data("Cory",102,22)

Details of:  Bill
Empoyee Id :  101
Age :  21
Company :  Samsung.com
-----------------------
Details of:  Cory
Empoyee Id :  102
Age :  22
Company :  Baidu.com


### Python Keyword Arguments

Keyword arguments are related to the function calls. A keyword argument is an argument value, passed to function preceded by the variable name and an equals sign.

This allows you to skip arguments or place them out of order because the Python interpreter is able to use the keywords provided to match the values with parameters.

In [11]:
# Example 1:

greet("Eric", "How do you do?")

# 2 keyword arguments
greet(name = "Eric",msg = "How do you do?")

# 2 keyword arguments (out of order)
greet(msg = "How do you do?",name = "Eric") 

# 1 positional, 1 keyword argument
greet("Eric", msg = "How do you do?")           

Hello Eric, How do you do?
Hello Eric, How do you do?
Hello Eric, How do you do?
Hello Eric, How do you do?


As we can see, we can mix positional arguments with keyword arguments during a function call. But we must keep in mind that keyword arguments must follow positional arguments.

Having a positional argument after keyword arguments will result in errors. For example, the function call as follows:

In [12]:
greet("Eric","How do you do?")

Hello Eric, How do you do?


In [13]:
greet(name="Eric","How do you do?")  # Will result in an error

SyntaxError: positional argument follows keyword argument (<ipython-input-13-0d17a834dca2>, line 1)

In [14]:
# Example 2:

def print_fullname(firstname, lastname):
    space = ' '
    full_name = firstname  + space + lastname
    return full_name
print(print_fullname(firstname = 'Milaan', lastname = 'Parmar'))

Milaan Parmar


In [15]:
# Example 3:

def add_two_numbers (num1, num2):
    total = num1 + num2
    return total
print(add_two_numbers(num2 = 3, num1 = 2)) # Order does not matter

5


In [16]:
# Example 4:

def calculate_age (current_year, birth_year):
    age = current_year - birth_year
    return age;
print('Age: ', calculate_age(2019, 1819))

Age:  200


In [17]:
# Example 5:

def remainder(dividend,divisor):
    x=dividend%divisor
    return x

#rem = remainder(10,3)
rem = remainder(divisor = 3, dividend = 10) # keyword argument
print("remainder of 10/3 : ",rem)

remainder of 10/3 :  1


In [18]:
# Example 6:

def message(name, surname):
    print("Hello", name, surname)

message(name="Frank", surname="Harper")
message(surname="Clayton", name="Gretta")

Hello Frank Harper
Hello Gretta Clayton


In keyword arguments order of argument is not matter, but the number of arguments must match. Otherwise, we will get an error.

While using keyword and positional argument simultaneously, we need to pass 1st arguments as positional arguments and then keyword arguments. Otherwise, we will get **`SyntaxError`**. See the following example.

In [19]:
# Example 7:

def message(first_nm, last_nm):
    print("Hello..!", first_nm, last_nm)

# correct use
message("Frank", "Harper")
message("Frank", last_nm="Harper")

message(first_nm="Frank", "Harper")  # SyntaxError: positional argument follows keyword argument

SyntaxError: positional argument follows keyword argument (<ipython-input-19-b8ab82a1062c>, line 10)

### Python Arbitrary/Variable-length Arguments

Sometimes, we do not know in advance the number of arguments that will be passed into a function. Python allows us to handle this kind of situation through function calls with an arbitrary number of arguments.

We can pass any number of arguments to this function. Internally all these values are represented in the form of a **tuple**.

In the function definition, we use an asterisk **`*`** before the parameter name to denote this kind of argument. For example:

In [20]:
# Example 1: 

def greet(*names):
    """This function greets all the person in the names tuple."""
    # names is a tuple with arguments
    for name in names:
        print("Hello", name)

greet("Gary", "Hank", "Ivan", "John")

Hello Gary
Hello Hank
Hello Ivan
Hello John


Here, we have called the function with multiple arguments. These arguments get wrapped up into a tuple before being passed into the function. Inside the function, we use a **`for`** loop to retrieve all the arguments back.

In [21]:
# Example 2: 

def addition(*numbers):
    total = 0
    for no in numbers:
        total = total + no
    print("Sum is:", total)

addition()                 # 0 arguments
addition(10, 5, 3, 6, 9)   # 5 arguments
addition(96, 77, 3.6)      # 3 arguments

Sum is: 0
Sum is: 33
Sum is: 176.6


In [22]:
# Example 3: 

def sum_all_nums(*nums):
    total = 0
    for num in nums:
        total += num     # same as total = total + num 
    return total
print(sum_all_nums(2, 3, 5)) # 10

10


Default and Arbitrary Number of Parameters in Functions

In [23]:
# Example 1: 

def printinfo( arg1, *vartuple ):
    "This prints a variable passed arguments"
    print ("Output is: ")
    print (arg1)
    for var in vartuple:
        print (var)
    return

    # Now you can call printinfo function
printinfo( 10 ) # We have given only one value for the argument
printinfo( 70, 40, 20 )
printinfo( 30, 60, 90, 120 )

Output is: 
10
Output is: 
70
40
20
Output is: 
30
60
90
120


In [24]:
# Example 2: 

def generate_groups (team,*args):
    print(team)
    for i in args:
        print(i)
print(generate_groups('Team-1','Milaan','Arthur','Clark','Ethan'))

Team-1
Milaan
Arthur
Clark
Ethan
None


## Function as a Parameter of Another Function

In [25]:
# Example 1: You can pass functions around as parameters

def square_number (n):
    return n * n
def do_something(f, x):
    return f(x)
print(do_something(square_number, 3)) # 27

9


# Python Recursion

In this class, you will learn to create a recursive function (a function that calls itself again and again).

# What is recursion?

Recursion is the process of defining something in terms of itself.

A physical world example would be to **place two parallel mirrors facing each other**. Any object in between them would be reflected recursively.

In [1]:
# Example 1:

def factorial(n):
    """This is a recursive function to find the factorial of an integer"""

    if n == 1:
        return 1
    else:
        return (n * factorial(n-1))  # 3 * 2 * 1 = 6


num = 3
print("The factorial of", num, "is", factorial(num))

The factorial of 3 is 6


In the above example, **`factorial()`** is a recursive function as it calls itself.

When we call this function with a positive integer, it will recursively call itself by decreasing the number.

Each function multiplies the number with the factorial of the number below it until it is equal to one. This recursive call can be explained in the following steps.

```python
factorial(3)          # 1st call with 3
3 * factorial(2)      # 2nd call with 2
3 * 2 * factorial(1)  # 3rd call with 1
3 * 2 * 1             # return from 3rd call as number=1
3 * 2                 # return from 2nd call
6                     # return from 1st call
```

Let's look at an image that shows a step-by-step process of what is going on:

<div>
<img src="../img/re2.png" width="350"/>
</div>

Our recursion ends when the number reduces to 1. This is called the base condition.

Every recursive function must have a base condition that stops the recursion or else the function calls itself infinitely.

The Python interpreter limits the depths of recursion to help avoid infinite recursions, resulting in stack overflows.

By default, the maximum depth of recursion is **`1000`**. If the limit is crossed, it results in **`RecursionError`**. Let's look at one such condition.

In [2]:
def recursor():
    recursor()
recursor()

RecursionError: maximum recursion depth exceeded

## Advantages of Recursion

1. Recursive functions make the code look clean and elegant.
2. A complex task can be broken down into simpler sub-problems using recursion.
3. Sequence generation is easier with recursion than using some nested iteration.

## Disadvantages of Recursion

1. Sometimes the logic behind recursion is hard to follow through.
2. Recursive calls are expensive (inefficient) as they take up a lot of memory and time.
3. Recursive functions are hard to debug.

# Python Anonymous/**`lambda`** Function

In this class, you'll learn about the anonymous function, also known as **`lambda`** functions. You'll learn what they are, their syntax and how to use them (with examples).

Let’s see an example to print even numbers without a **`lambda`** function and with a **`lambda`** function. See the difference in line of code as well as readability of code.

In [1]:
# Example 1: Program for even numbers without lambda function

def even_numbers(nums):
    even_list = []
    for n in nums:
        if n % 2 == 0:
            even_list.append(n)
    return even_list

num_list = [10, 9, 16, 78, 2, 3, 7, 1]
ans = even_numbers(num_list)
print("Even numbers are:", ans)

Even numbers are: [10, 16, 78, 2]


In [2]:
# Example 1: Program for even number with a lambda function

l = [10, 9, 16, 78, 2, 3, 7, 1]
even_nos = list(filter(lambda x: x % 2 == 0, l))
print("Even numbers are: ", even_nos)

Even numbers are:  [10, 16, 78, 2]


In [3]:
# Example 2: Program to show the use of lambda functions

double = lambda x: x * 2

print(double(6))

12


**Explanation:**

In the above program, lambda **`x: x * 2`** is the lambda function. Here **`x`** is the argument and **`x * 2`** is the expression that gets evaluated and returned.

This function has no name. It returns a function object which is assigned to the identifier **`double`**. We can now call it as a normal function. The statement

```python
>>> double = lambda x: x * 2
```

is nearly the same as:

```python
>>> def double(x):
>>>     return x * 2
```

In [4]:
# Example 3: Normal Function definition is here

def square(x):
    return x*x

# anonymous function
sqr = lambda x: x*x

#Calling square function
print("Square of number is",square(10)) #call normal function
print("Square of number is",sqr(2)) #call anonymous function

Square of number is 100
Square of number is 4


### `lambda` function use with `filter()`

The **`filter()`** function in Python takes in a function and a list as arguments.

The function is called with all the items in the list and a new list is returned which contains items for which the function evaluates to **`True`**.

Here is an example use of **`filter()`** function to filter out only even numbers from a list.

In [5]:
# Example 1: Program to filter out only the even items from a list

my_list = [1, 5, 4, 6, 8, 11, 3, 12]  # total 8 elements

new_list = list(filter(lambda x: (x%2 == 0), my_list)) # returns the output in form of a list

print("Even numbers are: ", new_list)

Even numbers are:  [4, 6, 8, 12]


In [6]:
# Example 2: 

l = [-10, 5, 12, -78, 6, -1, -7, 9]
positive_nos = list(filter(lambda x: x > 0, l))
print("Positive numbers are: ", positive_nos)

Positive numbers are:  [5, 12, 6, 9]


### `lambda` function with `map()`

The **`map()`** function in Python takes in a function and a list.

The function is called with all the items in the list and a new list is returned which contains items returned by that function for each item.

Here is an example use of **`map()`** function to double all the items in a list.

In [7]:
# Example 1: Program to double each item in a list using map()

my_list = [1, 5, 4, 6, 8, 11, 3, 12]  # total 8 elements
new_list = list(map(lambda x: x * 2, my_list)) # returns the output in form of a list
print("Double values are: ", new_list)

Double values are:  [2, 10, 8, 12, 16, 22, 6, 24]


In [8]:
# Example 2:

list1 = [2, 3, 4, 8, 9]
list2 = list(map(lambda x: x*x*x, list1))
print("Cube values are:", list2)

Cube values are: [8, 27, 64, 512, 729]


### `lambda` function with `reduce()`

The **`reduce()`** function is used to minimize sequence elements into a single value by applying the specified condition.

The **`reduce()`** function is present in the **`functools`** module; hence, we need to import it using the import statement before using it.

In [9]:
# Example 1:

from functools import reduce
list1 = [20, 13, 4, 8, 9]
add = reduce(lambda x, y: x+y, list1)
print("Addition of all list elements is : ", add)

Addition of all list elements is :  54


# Python Modules

In this class, you will learn to create and import custom modules in Python. Also, you will find different techniques to import and use custom and built-in modules in Python.

## What are modules in Python?

Module containing a set of codes or a set of functions which can be included to an application. Modules refer to the Python file, which contains Python code like Python statements, classes, functions, variables, etc. A file with Python code is defined with extension**`.py`**

For example: In **`main.py`**, where the **`main`** is the module name.

In Python, large code is divided into small modules. The benefit of modules is, it provides a way to share reusable functions.

## Advantage

* **Reusability**: Module can be used in some other python code. Hence it provides the facility of code reusability.
* **Categorization**: Similar type of attributes can be placed in one module.

## How to import modules in Python?

We can import the definitions inside a module to another module or the interactive interpreter in Python.

We use the **`import`** keyword to do this. To import our previously defined module example, we type the following in the Python prompt.

```python
>>> import example
```
This does not import the names of the functions defined in **`example`** directly in the current symbol table. It only imports the module name **`example`** there.

Using the module name we can access the function using the dot **`.`** operator. For example:

**# Example 1:**

In [1]:
import example
example.add(3,6.6)

9.6

Python has tons of standard modules. You can check out the full list of **[Python standard modules](https://docs.python.org/3/py-modindex.html)** and their use cases. These files are in the Lib directory inside the location where you installed Python.

Standard modules can be imported the same way as we import our user-defined modules.

There are various ways to import modules. They are listed below:

**# Example 2:**

The Python code for a module named a name normally resides in a file **main.py**. But first create a function in **test.py**. Here is an example of a simple module,

**test.py** - 

```python
>>> def fun():
>>>     print("something here inside fun()")
```
**main.py** −

```python
>>> import test
>>> from test import fun
>>> fun()
```

Output of above example: Run **main.py**

`something here inside fun()`

----------------------------------------------------------------
For **test.ipynb** or **test1.ipynb** and **main.ipynb**

Open anaconda prompt and type **`pip install import-ipynb`**

see:
Downloading import-ipynb-0.1.3.tar.gz (4.0 kB)

**test.py** or **test1.ipynb** - 

```python
>>> def fun():
>>>     print("something here inside fun()")
```
**main.ipynb** −

```python
>>> import import_ipynb

>>> import test1
>>> from test1 import fun
>>> fun()
```

Output of above example: Run **main.ipynb**

`something here inside fun()`

**# Example 3:**

**fibo.py** - 

```python
# Fibonacci numbers module
>>> def fib(n): # return Fibonacci series up to n
>>>     result = []
>>>     a, b = 0, 1
>>>     while b < n:
>>>         result.append(b)
>>>         a, b = b, a + b
>>>     return result
```
**main.py** −

```python
>>> import fibo
>>> from fibo import fib
>>> print(fib(100))
```

Output of above example: Run **main.py**

`[1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89]`

----------------------------------------------------------------
For **test.ipynb** or **test1.ipynb** and **main.ipynb**

Open anaconda prompt and type **`pip install import-ipynb`**

see:
Downloading import-ipynb-0.1.3.tar.gz (4.0 kB)

**fibo1.ipynb** or **fibo.py** - 

```python
# Fibonacci numbers module
>>> def fib(n): # return Fibonacci series up to n
>>>     result = []
>>>     a, b = 0, 1
>>>     while b < n:
>>>         result.append(b)
>>>         a, b = b, a + b
>>>     return result
```
**main.ipynb** −

```python
>>> import import_ipynb

>>> import fibo1
>>> from fibo1 import fib
>>> print(fib(100))
```

Output of above example: Run **main.ipynb**

`[1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89]`

**# Example 4:**

**equal.py** - 

```python
#function definition
>>> def equal(a, b):
>>>     if(a == b):
>>>         return True
```
**main.py** −

```python
>>> from equal import *
>>> a, b = 10, 10
>>> print(equal(a, b))
```

Output of above example: Run **main.py**

`True`

----------------------------------------------------------------
For **equal.ipynb** or **equal1.ipynb** and **main.ipynb**

Open anaconda prompt and type **`pip install import-ipynb`**

see:
Downloading import-ipynb-0.1.3.tar.gz (4.0 kB)

**equal.ipynb** or **equal1.ipynb** - 

```python
#function definition
>>> def equal(a, b):
>>>     if(a == b):
>>>         return True
```
**main.ipynb** −

```python
>>> import import_ipynb

>>> from equal1 import *
>>> a, b = 10, 10
>>> print(equal(a, b))
```

Output of above example: Run **main.ipynb**

`True`

In Python, each and every built-in module has a large number of predefined functions to perform specific tasks. For instance, Python has a built-in module named **`operator`**, inside which the function called **`eq()`** is defined. This function returns the boolean value of **`True`** if the given two input values are equal. Else, returns **`False`**.

So now, we can make use of this operator module in our **`main()`** program to check if two numbers are equal. This means we would no longer need that **`equal.py`** module.

In [2]:
# Example 5: 

#main function
from operator import *
a, b = 10, 10
print(eq(a, b))

True


### Python `import` statement

We can import a module using the **`import`** statement and access the definitions inside it using the dot operator **`.`** as described above. Here is an example.

In [3]:
# Example 1: 

# import statement example to import standard module math
import math
print("The value of pi is", math.pi)

The value of pi is 3.141592653589793


In [4]:
# Example 2:

import math

# use math module functions
print(math.sqrt(6))    # Output 2.449489742783178

2.449489742783178


### `import` multiple modules

If we want to use more than one module, then we can import multiple modules. This is the simplest form of **`import`** a keyword that we already use in the above example.

In [5]:
# Example 1: Import two modules

import math, random

print(math.factorial(6))
print(random.randint(10, 30))

720
28


### Python `from-import` statement

To import particular classes or functions, we can use the **`from-import`** statement. It is an alternate way to **`import`**. By using this form, we can import individual attributes and methods directly into the program.We can import specific names from a module without importing the module as a whole.

In [6]:
# Example 1: import only factorial function from math module
from math import factorial

print(factorial(6))

720


In [7]:
# Example 2: 

from math import pi, e
e

2.718281828459045

In [8]:
# Example 3: 

# import only pi from math module
from math import pi
print("The value of pi is", pi)

The value of pi is 3.141592653589793


Here, we imported only the **`pi`** attribute from the **`math`** module.

In such cases, we don't use the dot operator. We can also import multiple attributes as follows:

In [9]:
# Example 4: 

from math import pi, e
pi

3.141592653589793

### `import` with renaming

We can import a module by renaming it as follows:

In [10]:
# Example 1: 

# import module by renaming it
import math as m
print("The value of pi is", m.pi)

The value of pi is 3.141592653589793


We have renamed the **`math`** module as **`m`**. This can save us typing time in some cases.

>**Note:** that the name **`math`** is not recognized in our scope. Hence, **`math.pi`** is invalid, and **`m.pi`** is the correct implementation.

In [11]:
# Example 2:  Import a module by renaming it

import random as rand

print(rand.randrange(10, 20, 2))

12


In [12]:
# Example 3:  import a method by renaming it

# rename randint as random_number
from random import randint as random_number

# Gives any random number from range(10, 50)
print(random_number(30, 60))

49


### `import` all names

We can **`import`** all functions and attributes of a specific module, then instead of writing all function names and attribute names, we can import all using an **asterisk** **`*`**.

In [13]:
# Example 1: 

# import all names from the standard module math

from math import *
print("The value of pi is", pi)

The value of pi is 3.141592653589793


Here, we have imported all the definitions from the **`math`** module. This includes all names visible in our scope except those beginning with an underscore(private definitions).

Importing everything with the asterisk **`*`** symbol is not a good programming practice. This can lead to duplicate definitions for an identifier. It also hampers the readability of our code.

In [14]:
# Example 2: 

from math import *
print(pow(6,3))
print(factorial(6))

print(pi*3)
print(sqrt(100))

216.0
720
9.42477796076938
10.0


## Python Module Search Path

While importing a module, Python looks at several places. Interpreter first looks for a built-in module. Then(if built-in module not found), Python looks into a list of directories defined in **`sys.path`**. The search is in this order.

1. The current directory.
2. **`PYTHONPATH`** (an environment variable with a list of directories).
3. The installation-dependent default directory.

In [15]:
import sys
sys.path

['C:\\Users\\Deepak\\01_Learn_Python4Data\\04_Python_Functions',
 'C:\\ProgramData\\Anaconda3\\python38.zip',
 'C:\\ProgramData\\Anaconda3\\DLLs',
 'C:\\ProgramData\\Anaconda3\\lib',
 'C:\\ProgramData\\Anaconda3',
 '',
 'C:\\Users\\Deepak\\AppData\\Roaming\\Python\\Python38\\site-packages',
 'C:\\ProgramData\\Anaconda3\\lib\\site-packages',
 'C:\\ProgramData\\Anaconda3\\lib\\site-packages\\locket-0.2.1-py3.8.egg',
 'C:\\ProgramData\\Anaconda3\\lib\\site-packages\\win32',
 'C:\\ProgramData\\Anaconda3\\lib\\site-packages\\win32\\lib',
 'C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Pythonwin',
 'C:\\ProgramData\\Anaconda3\\lib\\site-packages\\IPython\\extensions',
 'C:\\Users\\Deepak\\.ipython']

We can add and modify this list to add our own path.

## Reloading a module

The Python interpreter imports a module only once during a session. This makes things more efficient. Foe example:

**# Example 1:**

Suppose we have the following code in a module named **`my_module`**.

```python
# This module shows the effect of
#  multiple imports and reload

print("This code got executed")
```

In [16]:
import my_module

This code got executed


We can see that our code got executed only once. This goes to say that our module was imported only once.

Now if our module changed during the course of the program, we would have to reload it.One way to do this is to restart the interpreter. But this does not help much.

Python provides a more efficient way of doing this. We can use the **`reload()`** function inside the **`imp`** module to reload a module. We can do it in the following ways:

In [17]:
import imp
import my_module

In [18]:
import my_module
imp.reload(my_module)

This code got executed


<module 'my_module' from 'C:\\Users\\Deepak\\01_Learn_Python4Data\\04_Python_Functions\\my_module.py'>

**# Example 2:**

First, create a Python module with the name **`greet_module.pynb`**
and write the below code in that file.

```python
>>> print("Welcome to Dr Parmar's Python4DataScience class")
```

Now, create a Python file with the name, **main.py** and write the below code in it and import the module test_module.py. See the following code.

```python
>>> import import_ipynb
>>> import time
>>> from importlib import reload

>>> # load 1st time
>>> import test_module
>>> time.sleep(25)
>>> # reload 
>>> reload(test_module)
>>> time.sleep(25) 

>>> # reload again  
>>> reload(test_module)
>>> print("This is test file..")
```

Output of above example: Run **main.py** 

```python
Welcome to Dr Parmar's Python4DataScience class
Welcome to Dr Parmar's Python4DataScience class
Welcome to Dr Parmar's Python4DataScience class
This is test file..
```

## The `dir()` built-in function

We can use the **`dir()`** function to find out names that are defined inside a module. When we use this function with any object (an object can be sequence like list, tuple, set, dict or can be class, function, module, etc. ), it returns properties, attributes, and method.

For example, we have defined a function **`add()`** in the module example that we had in the beginning.

We can use dir in example module in the following way:

In [19]:
# Example 1:

dir(example)

['__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'add']

Here, we can see a sorted list of names (along with **`add`**). All other names that begin with an underscore are default Python attributes associated with the module (not user-defined).

For example, the **`__name__`** attribute contains the name of the module.

In [20]:
# Example 2:

import example
example.__name__

'example'

All the names defined in our current namespace can be found out using the **`dir()`** function without any arguments.

In [21]:
# Example 3:

a = 1
b = "hello"
import math
dir()
['__builtins__', '__doc__', '__name__', 'a', 'b', 'math', 'pyscripter']

['__builtins__', '__doc__', '__name__', 'a', 'b', 'math', 'pyscripter']

In [22]:
# Example 4:

import math

print(dir(math))

['__doc__', '__loader__', '__name__', '__package__', '__spec__', 'acos', 'acosh', 'asin', 'asinh', 'atan', 'atan2', 'atanh', 'ceil', 'comb', 'copysign', 'cos', 'cosh', 'degrees', 'dist', 'e', 'erf', 'erfc', 'exp', 'expm1', 'fabs', 'factorial', 'floor', 'fmod', 'frexp', 'fsum', 'gamma', 'gcd', 'hypot', 'inf', 'isclose', 'isfinite', 'isinf', 'isnan', 'isqrt', 'ldexp', 'lgamma', 'log', 'log10', 'log1p', 'log2', 'modf', 'nan', 'perm', 'pi', 'pow', 'prod', 'radians', 'remainder', 'sin', 'sinh', 'sqrt', 'tan', 'tanh', 'tau', 'trunc']


### Example Using `os` Module

Using python **`os`** module it is possible to automatically perform many operating system tasks. The **`os`** module in Python provides functions for creating, changing current working directory, and removing a directory (folder), fetching its contents, changing and identifying the current directory.

In [23]:
# Example 1:

# import the module
import os
# Creating a directory
os.mkdir('directory_name')
# Changing the current directory
os.chdir('path')
# Getting current working directory
os.getcwd()
# Removing directory
os.rmdir()

FileExistsError: [WinError 183] Cannot create a file when that file already exists: 'directory_name'

### Example Using `sys` Module

The **[sys](https://docs.python.org/3/library/sys.html)** module provides functions and variables used to manipulate different parts of the Python runtime environment.

**`sys.argv`** function returns a list of command line arguments passed to a Python script. The item at index 0 in this list is always the name of the script, at index 1 is the argument passed from the command line.

In [24]:
import sys
sys.path

['C:\\Users\\Deepak\\01_Learn_Python4Data\\04_Python_Functions',
 'C:\\ProgramData\\Anaconda3\\python38.zip',
 'C:\\ProgramData\\Anaconda3\\DLLs',
 'C:\\ProgramData\\Anaconda3\\lib',
 'C:\\ProgramData\\Anaconda3',
 '',
 'C:\\Users\\Deepak\\AppData\\Roaming\\Python\\Python38\\site-packages',
 'C:\\ProgramData\\Anaconda3\\lib\\site-packages',
 'C:\\ProgramData\\Anaconda3\\lib\\site-packages\\locket-0.2.1-py3.8.egg',
 'C:\\ProgramData\\Anaconda3\\lib\\site-packages\\win32',
 'C:\\ProgramData\\Anaconda3\\lib\\site-packages\\win32\\lib',
 'C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Pythonwin',
 'C:\\ProgramData\\Anaconda3\\lib\\site-packages\\IPython\\extensions',
 'C:\\Users\\Deepak\\.ipython']

**First**, create a Python module with the name **`script.py`** and write the below code in that file.

```py
>>> import sys
>>> # print(sys.argv[0], argv[1],sys.argv[2])  # this line would print out: filename argument1 argument2
>>> print('Welcome {}. Enjoy  {} for Data Science ^_^'.format(sys.argv[1], sys.argv[2]))
```
**Second**, open **`Anaconda Prompt (Anaconda3)`**

**Third**, type system path after **`cd`**

```sy
cd C:\\Users\\Deepak\\01_Learn_Python4Data\\04_Python_Functions
```

**Forth**, type

```sy
python script.py Milaan Python
```

**Output:**

```sy
Welcome Milaan. Enjoy  Python for Data Science ^_^
```

<div>
<img src="../img/anaconda_prompt.png" width="1000"/>
</div>

Some useful **`sys`** commands to type in:

```py
# to exit sys
sys.exit()
# To know the largest integer variable it takes
sys.maxsize
# To know environment path
sys.path
# To know the version of python you are using
sys.version
```

### Example Using `operator` Module

In [25]:
# Example 1:

from operator import *
a, b = 10, 20
#prints the product of the values 'a' and 'b'
print(mul(a, b))
#prints True if the value of 'a' is greater than 'b'. Else, False
print(gt(a, b))
#prints the remainder value, when the value of 'a' is divided by 'b'
print(mod(a, b))
#concatenates and prints the given two strings
print(concat("Hello ", "World"))

200
False
10
Hello World


In [26]:
# Example 1:

from string import *
print(capwords("hello world")) #capitalizes the first letter of each words
print(ascii_letters) #prints all lowercase and uppercase letters

Hello World
abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ


In [27]:
# Example 2:

import string
print(string.ascii_letters) # abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ
print(string.digits)        # 0123456789
print(string.punctuation)   # !"#$%&'()*+,-./:;<=>?@[\]^_`{|}~

abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ
0123456789
!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


### Example Using `decimal` Module

In [28]:
# Example 1:

from decimal import *
a, b = 10, 3
c = a / b
print(c)
print(Decimal(c)) #prints the complete decimal value of c

3.3333333333333335
3.333333333333333481363069950020872056484222412109375


In [29]:
# Example 1:

from random import *
print(randint(10, 20)) #prints a random number between the given range

list1 = [30, 23, 45, 16, 89, 56]
print(choice(list1)) #prints a random element from the given iterator

print(uniform(10, 20)) #prints a random float number between two given values

10
16
13.03505598723366


In [30]:
# Example 2:

import random

print(random.random())
print(random.randint(2,8))

0.2946014910624595
8


In [31]:
# Example 3:

from random import randint   # import randint function

print(random())   # it doesn't take any arguments; it returns a value between 0 and 0.9999

# call randint function to get random number
print(randint(10, 30))  # it returns a random integer number between [10, 30] inclusive

TypeError: 'module' object is not callable

In [32]:
# Example 1:

from math import *
print(sqrt(16)) # prints the square root of the value 16 in the form of a floating-point value
print(factorial(5)) # prints the factorial of the value 5

4.0
120


In [33]:
# Example 2:

import math

a=4.6
print(math.ceil(a))       # 5, rounding to the highest
print(math.floor(a))      # 4, rounding to the lowest
b=9
print(math.sqrt(b))       # 3.0, square root
print(math.exp(3.0))      # 20.085536923187668
print(math.log(2.0))      # 0.6931471805599453
print(math.pow(2.0,3.0))  # 8.0, exponential function
print(math.sin(0))        # 0.0
print(math.cos(0))        # 1.0
print (math.tan(45))      # 1.6197751905438615
print(math.log10(100))    # 2.0, logarithm with 10 as base 

5
4
3.0
20.085536923187668
0.6931471805599453
8.0
0.0
1.0
1.6197751905438615
2.0


#### Constants  -  

The math module provides two constants for mathematical Operations:

* **`math.pi`** : Returns constant Pi = 3.14159...
* **`math.e`** : Returns constant e= 2.71828...

In [34]:
# Example 3:

import math 

print("math.pi : ",math.pi)  # 3.141592653589793, pi constant
print("math.e : ",math.e)    # 2.718281828459045, e  constant

math.pi :  3.141592653589793
math.e :  2.718281828459045


### Example Using `statistics` Module

The statistics module provides functions for mathematical statistics of numeric data. The popular statistical functions which are defined in this module: **`mean`**, **`median`**, **`mode`**, **`stdev`** etc.

In [35]:
# Example 1:

from statistics import * # importing all the statistics modules
ages = [20, 20, 4, 24, 25, 22, 26, 20, 23, 22, 26]
print(mean(ages))       # ~22.9
print(median(ages))     # 23
print(mode(ages))       # 20
print(stdev(ages))      # ~2.3

21.09090909090909
22
20
6.106628291529549


### Wrap up

Since Python provides a lot of built-in modules, it is advisable to use built-in modules rather than user-created modules to perform basic operations.

## 💻 Exercises ➞ <span class='label label-default'>Module</span>

### Exercises ➞ <span class='label label-default'>Level 1</span>

1. Writ a function which generates a six digit/character **`random_user_id`**.
    - ```py
     print(random_user_id());
     '1ee33d'
    ```
2. Modify the previous task. Declare a function named **`user_id_gen_by_user`**. It doesn’t take any parameters but it takes two inputs using **`input()`**. One of the inputs is the number of characters and the second input is the number of IDs which are supposed to be generated.
   
    - ```py
print(user_id_gen_by_user()) # user input: 5 5
#output:
#kcsy2
#SMFYb
#bWmeq
#ZXOYh
#2Rgxf
    ``` 
 
    - ```py
print(user_id_gen_by_user()) # 16 5
#1GCSgPLMaBAVQZ26
#YD7eFwNQKNs7qXaT
#ycArC5yrRupyG00S
#UbGxOFI7UXSWAyKN
#dIV0SSUTgAdKwStr
    ```

3. Write a function named **`rgb_color_gen`**. It will generate rgb colors (3 values ranging from 0 to 255 each).
   
    - ```py
print(rgb_color_gen()) 
#rgb(125,244,255) - the output should be in this form
    ```
 
### Exercises ➞ <span class='label label-default'>Level 2</span>

1. Write a function **`list_of_hexa_colors`** which returns any number of hexadecimal colors in an array (six hexadecimal numbers written after **`#`**. Hexadecimal numeral system is made out of 16 symbols, 0-9 and first 6 letters of the alphabet, a-f. Check the task 6 for output examples).
2. Write a function **`list_of_rgb_colors`** which returns any number of RGB colors in an array.
3. Write a function **`generate_colors`** which can generate any number of hexa or rgb colors.

    - ```py
   generate_colors('hexa', 3) # ['#a3e12f','#03ed55','#eb3d2b'] 
   generate_colors('hexa', 1) # ['#b334ef']
   generate_colors('rgb', 3)  # ['rgb(5, 55, 175','rgb(50, 105, 100','rgb(15, 26, 80'] 
   generate_colors('rgb', 1)  # ['rgb(33,79, 176)']
    ```
   
### Exercises ➞ <span class='label label-default'>Level 3</span>

1. Call your function **`shuffle_list`**, it takes a list as a parameter and it returns a shuffled list
2. Write a function which returns an array of seven random numbers in a range of 0-9. All the numbers must be unique.

# Python Random Module

You can generate random numbers in Python by using random module.

Python offers **`random`** module that can generate random numbers.

These are pseudo-random number as the sequence of number generated depends on the seed.

If the seeding value is same, the sequence will be the same. For example, if you use 2 as the seeding value, you will always see the following sequence.

In [1]:
import random
random.seed(2)

print(random.random())
print(random.random())
print(random.random())

0.9560342718892494
0.9478274870593494
0.05655136772680869


**Not so random eh?** Since this generator is completely deterministic, it must not be used for encryption purpose.

Here is the list of all the functions defined in random module with a brief explanation of what they do.

**List of Functions in Python Random Module**

| Function | Description |
|:----| :--- |
| **`seed(a=None, version=2)`** | Initialize the random number generator | 
| **`getstate()`** | Returns an object capturing the current internal state of the generator | 
| **`setstate(state)`** | Restores the internal state of the generator | 
| **`getrandbits(k)`** | Returns a Python integer with k random bits | 
| **`randrange(start, stop[, step])`** | Returns a random integer from the range | 
| **`randint(a, b)`** | Returns a random integer between a and b inclusive | 
| **`choice(seq)`** | Return a random element from the non-empty sequence | 
| **`shuffle(seq)`** | Shuffle the sequence | 
| **`sample(population, k)`** | Return a k length list of unique elements chosen from the population sequence | 
| **`random()`** | Return the next random floating point number in the range [0.0, 1.0] | 
| **`uniform(a, b)`** | Return a random floating point number between a and b inclusive | 
| **`triangular(low, high, mode)`** | Return a random floating point number between low and high, with the specified mode between those bounds | 
| **`betavariate(alpha, beta)`** | Beta distribution | 
| **`expovariate(lambd)`** | Exponential distribution | 
| **`gammavariate(alpha, beta)`** | Gamma distribution | 
| **`gauss(mu, sigma)`** | Gaussian distribution | 
| **`lognormvariate(mu, sigma)`** | Log normal distribution | 
| **`normalvariate(mu, sigma)`** | Normal distribution | 
| **`vonmisesvariate(mu, kappa)`** | Vonmises distribution | 
| **`paretovariate(alpha)`** | Pareto distribution | 
| **`weibullvariate(alpha, beta)`** | Weibull distribution | 

Visit this page to learn more on **[how you can generate pseudo-random numbers in Python](https://docs.python.org/3/library/random.html)**.

# Python Mathematical Functions

Learn about all the mathematical functions available in Python and how you can use them in your program.

In [1]:
# Square root calculation

import math
math.sqrt(4)

2.0

This module does not support complex datatypes. The **[cmath module](https://docs.python.org/3.0/library/cmath.html)** is the **`complex`** counterpart.

## Functions in Python Math Module

Here is the list of all the functions and attributes defined in **`math`** module with a brief explanation of what they do.

**List of Functions in Python Math Module**

| Function | Description |
|:----| :--- |
| **`ceil(x)`** | Returns the smallest integer greater than or equal to x. | 
| **`copysign(x, y)`** | Returns x with the sign of y | 
| **`fabs(x)`** | Returns the absolute value of x | 
| **`factorial(x)`** | Returns the factorial of x | 
| **`floor(x)`** | Returns the largest integer less than or equal to x | 
| **`fmod(x, y)`** | Returns the remainder when x is divided by y | 
| **`frexp(x)`** | Returns the mantissa and exponent of x as the pair (m, e) | 
| **`fsum(iterable)`** | Returns an accurate floating point sum of values in the iterable | 
| **`isfinite(x)`** | Returns True if x is neither an infinity nor a NaN (Not a Number) | 
| **`isinf(x)`** | Returns True if x is a positive or negative infinity | 
| **`isnan(x)`** | Returns True if x is a NaN | 
| **`ldexp(x, i)`** | Returns x * (2**i) | 
| **`modf(x)`** | Returns the fractional and integer parts of x | 
| **`trunc(x)`** | Returns the truncated integer value of x | 
| **`exp(x)`** | Returns e**x | 
| **`expm1(x)`** | Returns e**x - 1 | 
| **`log(x[, b])`** | Returns the logarithm of **`x`** to the base **`b`** (defaults to e) | 
| **`log1p(x)`** | Returns the natural logarithm of 1+x | 
| **`log2(x)`** | Returns the base-2 logarithm of x | 
| **`log10(x)`** | Returns the base-10 logarithm of x | 
| **`pow(x, y)`** | Returns x raised to the power y | 
| **`sqrt(x)`** | Returns the square root of x | 
| **`acos(x)`** | Returns the arc cosine of x | 
| **`asin(x)`** | Returns the arc sine of x | 
| **`atan(x)`** | Returns the arc tangent of x | 
| **`atan2(y, x)`** | Returns atan(y / x) | 
| **`cos(x)`** | Returns the cosine of x | 
| **`hypot(x, y)`** | Returns the Euclidean norm, sqrt(x*x + y*y) | 
| **`sin(x)`** | Returns the sine of x | 
| **`tan(x)`** | Returns the tangent of x | 
| **`degrees(x)`** | Converts angle x from radians to degrees | 
| **`radians(x)`** | Converts angle x from degrees to radians | 
| **`acosh(x)`** | Returns the inverse hyperbolic cosine of x | 
| **`asinh(x)`** | Returns the inverse hyperbolic sine of x | 
| **`atanh(x)`** | Returns the inverse hyperbolic tangent of x | 
| **`cosh(x)`** | Returns the hyperbolic cosine of x | 
| **`sinh(x)`** | Returns the hyperbolic cosine of x | 
| **`tanh(x)`** | Returns the hyperbolic tangent of x | 
| **`erf(x)`** | Returns the error function at x | 
| **`erfc(x)`** | Returns the complementary error function at x | 
| **`gamma(x)`** | Returns the Gamma function at x | 
| **`lgamma(x)`** | Returns the natural logarithm of the absolute value of the Gamma function at x | 
| **`pi`** | Mathematical constant, the ratio of circumference of a circle to it's diameter (3.14159...) | 
| **`e`** | mathematical constant e (2.71828...) | 

Visit this page to learn about all the **[mathematical functions defined in Python 3](https://docs.python.org/3/library/math.html)**.

# Python Package

In this class, you'll learn to divide your code base into clean, efficient modules using Python packages. Also, you'll learn to import and use your own or third party packagesin your Python program.